### PR TITLE
plan: use the inferred type as the column type in the schema

### LINKED ISSUE #7307: server crash on GROUP BY indexed enum
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

```
DROP TABLE IF EXISTS t1;

CREATE TABLE t1 (
 id INT NOT NULL PRIMARY KEY auto_increment,
 b varchar(255) NOT NULL,
 c ENUM('a', 'b', 'c', 'd', 'e'),
 INDEX (c)
);

INSERT INTO t1 VALUES 
 (NULL, REPEAT('b', 255), 'a'),
 (NULL, REPEAT('b', 255), 'b'),
 (NULL, REPEAT('b', 255), 'c'),
 (NULL, REPEAT('b', 255), 'd'),
 (NULL, REPEAT('b', 255), 'e');
 
SELECT c FROM t1 GROUP BY c;

```

2. What did you expect to see?

Not crash.

3. What did you see instead?

```
2018/08/07 14:43:43.857 conn.go:427: [error] lastCmd SELECT c FROM t1 GROUP BY c, runtime error: slice bounds out of range, goroutine 38207 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc42087b2b0, 0xc4209f5dff)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:425 +0x10e
panic(0x1155280, 0x1e09700)
        /usr/local/go/src/runtime/panic.go:502 +0x229
github.com/pingcap/tidb/server.(*clientConn).writeResultset.func1(0x0, 0x1445c00, 0xc420b0f2c0, 0xc4209f5bf8, 0xc42087b2b0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:943 +0x325
panic(0x1155280, 0x1e09700)
        /usr/local/go/src/runtime/panic.go:502 +0x229
github.com/pingcap/tidb/util/chunk.Row.getNameValue(0xc420aaeca0, 0x0, 0x0, 0x400, 0x573d1b, 0xc42096602a)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/util/chunk/row.go:106 +0x126
github.com/pingcap/tidb/util/chunk.Row.GetEnum(0xc420aaeca0, 0x0, 0x0, 0xc4209f5978, 0x1, 0x9)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/util/chunk/row.go:112 +0x3f
github.com/pingcap/tidb/server.dumpTextRow(0xc420625000, 0x4, 0x400, 0xc421869c80, 0x1, 0x1, 0xc420aaeca0, 0x0, 0x50, 0x121a620, ...)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/util.go:326 +0xcfe
github.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc42087b2b0, 0x7f4649c81130, 0xc420f02d00, 0x1445c00, 0xc420b0f2c0, 0xc400005b00, 0xc42087b2b0, 0x0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:1015 +0x2c5
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc42087b2b0, 0x7f4649c81130, 0xc420f02d00, 0x1445c00, 0xc420b0f2c0, 0xc400007c00, 0x0, 0x0, 0x0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:956 +0x1bc
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc42087b2b0, 0x7f4649c81130, 0xc420f02d00, 0xc42003c681, 0x1b, 0x0, 0x0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:873 +0x124
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc42087b2b0, 0xc42003c681, 0x1c, 0x1c, 0x0, 0x0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:612 +0x4f9
github.com/pingcap/tidb/server.(*clientConn).Run(0xc42087b2b0)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:470 +0x1be
github.com/pingcap/tidb/server.(*Server).onConn(0xc4207168a0, 0x14478e0, 0xc421869888)
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:324 +0x22b
created by github.com/pingcap/tidb/server.(*Server).Run
        /home/jenkins/workspace/build_tidb_master/go/src/github.com/pingcap/tidb/server/server.go:264 +0x4dc

```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
MySQL [(none)]> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v2.1.0-beta-156-g5e7aa1d
Git Commit Hash: 5e7aa1d97d6459843949ae3587c9b5ac6661bb37
Git Branch: master
UTC Build Time: 2018-08-01 02:49:37
GoVersion: go version go1.10.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
1 row in set (0.00 sec)

```

### LINKED ISSUE #7578: slice bounds out of range
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
programer

2. What did you expect to see?
371535815214_ pic

3. What did you see instead?
381535815342_ pic
391535815362_ pic_hd

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
 Release Version: v2.1.0-rc.1-7-g38c939f

### PR DESCRIPTION
**Pull request by zz-jason**
### What problem does this PR solve? 

fix #7578 
fix #7307 

The type of the column in the schema should be reset after creating an aggregate function descriptor through the function `aggregation.NewAggFuncDesc()`

### Check List 

Tests 

 - **Unit test** WIP
 - Integration test

Related changes
 - Need to be included in the release note

### CODE REVIEW

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -594,6 +596,9 @@ func (b *planBuilder) buildDistinct(child LogicalPlan, length int) *LogicalAggre
 	}
 	plan4Agg.SetChildren(child)
 	plan4Agg.SetSchema(child.Schema().Clone())
+	for i, col := range plan4Agg.schema.Columns {
+		col.RetType = plan4Agg.AggFuncs[i].RetTp
+	}
 	// Distinct will be rewritten as first_row, we reset the type here since the return type
```
Comment (eurekaka) [inline plan/logical_plan_builder.go]: this function already has this type update in next few lines.

File: plan/rule_decorrelate.go [hunk 2]
```diff
@@ -182,9 +182,10 @@ func (s *decorrelateSolver) optimize(p LogicalPlan) (LogicalPlan, error) {
 				agg.SetSchema(apply.Schema())
 				agg.GroupByItems = expression.Column2Exprs(outerPlan.Schema().Keys[0])
 				newAggFuncs := make([]*aggregation.AggFuncDesc, 0, apply.Schema().Len())
-				for _, col := range outerPlan.Schema().Columns {
+				for i, col := range outerPlan.Schema().Columns {
 					first := aggregation.NewAggFuncDesc(agg.ctx, ast.AggFuncFirstRow, []expression.Expression{col}, false)
 					newAggFuncs = append(newAggFuncs, first)
+					outerPlan.Schema().Columns[i].RetType = first.RetTp
```
Comment (eurekaka) [inline plan/rule_decorrelate.go]: Is it safe to modify the child schema directly? should we make a clone here?

File: plan/logical_plan_builder.go [hunk 3]
```diff
@@ -115,7 +115,9 @@ func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.Aggrega
 	for _, col := range p.Schema().Columns {
 		newFunc := aggregation.NewAggFuncDesc(b.ctx, ast.AggFuncFirstRow, []expression.Expression{col}, false)
 		plan4Agg.AggFuncs = append(plan4Agg.AggFuncs, newFunc)
-		schema4Agg.Append(col)
+		newCol, _ := col.Clone().(*expression.Column)
```
Comment (winoros) [inline plan/logical_plan_builder.go]: So one column occurs in agg and agg' child. The tp of this two is different?

Comment (zz-jason) [reply plan/logical_plan_builder.go]: Possibly. Actually, they are two columns if the type inferred after `NewAggFuncDesc()` is different from the origin type.

**Comment by zz-jason** [comment]
@winoros @eurekaka @XuHuaiyu PTAL

File: plan/logical_plan_builder.go (hunk 3 above)
Comment (winoros) [reply plan/logical_plan_builder.go]: Since they are different, why `unique id` is the same one.

**Comment by winoros** [comment]
@zz-jason Should we let the columns have same `UniqueID` have different `fieldType`? Or should we call `Clone` before `ResolveIndices` anymore?

**Comment by zz-jason** [comment]
@winoros I think the `UniqueID` of a column should be carefully maintained case by case.

**Comment by winoros** [comment]
@zz-jason Should we keep the `firstRow` function just like other agg func instead of unwarping the `firstRow` so the schema needs to change the column's `fieldtype`?

**Comment by winoros** [comment]
I've figured out why new column cannot take effect easily. Will try to fix it when i have time.

**Pull request merged**